dhis-2/dhis-services/dhis-service-dxf2/src/main/java/org/hisp/dhis/dxf2/events/eventdatavalue/DefaultEventDataValueService.java

```diff
@@ -116,16 +119,32 @@ public void processDataValues( ProgramStageInstance programStageInstance, Event
             {
                 // This can happen if a wrong data element identifier is provided
                 importSummary.getConflicts().add( new ImportConflict( "dataElement", dataValue.getDataElement() + " is not a valid data element" ) );
+                continue;
             }
-            else if ( validateDataValue( programStageInstance, importOptions.getUser(), dataElement, dataValue.getValue(), importSummary )
-                && !importOptions.isDryRun() )
+
+            if ( validateDataValue( programStageInstance, importOptions.getUser(), dataElement, dataValue.getValue(), importSummary ) )
             {
+                accessibleDataElements.add( dataValue.getDataElement() );
                 prepareDataValueForStorage( dataElementValueMap, dataValue, dataElement, newDataValues, updatedDataValues,
                     removedDataValuesDueToEmptyValue, storedBy );
             }
+            else
+            {
+                nonAccessibleDataElements.add( dataValue.getDataElement() );
+            }
         }
 
-        programStageInstanceService.auditDataValuesChangesAndHandleFileDataValues( newDataValues, updatedDataValues, removedDataValuesDueToEmptyValue, dataElementsCache, programStageInstance, singleValue );
+        Set<String> checkedDataElements = Sets.union( accessibleDataElements, nonAccessibleDataElements );
+
+        for ( DataElement dataElement : programStageInstance.getProgramStage().getDataElements() )
+        {
+            if ( !checkedDataElements.contains( dataElement.getUid() ) &&  !trackerAccessManager.canWrite( importOptions.getUser(), programStageInstance, dataElement, true ).isEmpty() )
```

Comment:
the call to `trackerAccessManager.canWrite( importOptions.getUser(), programStageInstance, dataElement, true )` is taking place also at line 125 (inside `validateDataValue`). I understand that this new block of code (from line 137 to 145) is required for the auditing. Can we somehow avoid the call to this `canWrite` method a second time?

Reply:
we are not checking `canWrite`for the same object. otherwise we need to refactor the entire class, because the way it is written today focuses only on the incoming payload. what we are checking the second time (at least in the ACL) is those not in the payload.